### PR TITLE
[sgl-kernel] Support FlashInfer top_k_top_p_sampling_from_logits

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation

This PR is to introduce FlashInfer top_k_top_p_sampling_from_logits kernel into SGLang, which is expected to improve long context inference performance. E2E benchmark result shows slightly improvement in max_concurrency==1.

The principle of the op is to adjust the order of softmax and reduce the number of the elements to participant softmax.
logits -> top_k_mask_logits -> softmax -> top_p_sampling_from_probs
instead of:
logits -> softmax -> top_k_renorm_probs -> top_p_sampling_from_probs

Inspired by @bbuf. Thanks!

## Accuracy Tests
```
This PR in B200:
➜  sglang git:(main) ✗ python3 benchmark/gsm8k/bench_sglang.py --num-questions 200 --parallel 128 --num-shots 8 --port 30000
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:11<00:00, 17.67it/s]
Accuracy: 0.725
Invalid: 0.000
Latency: 11.433 s
Output throughput: 5060.150 token/s

Main in H20:
[root  /root/luoyuan.luo/workspace/SGLang] 二 8月 12 19:19:10 
$python3 benchmark/gsm8k/bench_sglang.py --num-questions 200 --parallel 128 --num-shots 8 --port 30000 --data-path /root/luoyuan.luo/test.jsonl 
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:11<00:00, 17.40it/s]
Accuracy: 0.705
Invalid: 0.000
Latency: 11.691 s
Output throughput: 4676.880 token/s

➜  sgl-kernel git:(flashinfer_topk_topp_softmax) python -m pytest ./tests/test_sampling.py
=============================================================================================================== test session starts ===============================================================================================================
platform linux -- Python 3.10.12, pytest-8.4.1, pluggy-1.6.0
rootdir: /sgl-workspace/sglang_dev2/sgl-kernel
configfile: pyproject.toml
plugins: typeguard-4.4.4, anyio-4.9.0
collected 153 items

tests/test_sampling.py ...................................................................................................sss...................................................                                                            [100%]

========================================================================================================= 150 passed, 3 skipped in 25.28s =========================================================================================================
```

## Benchmarking and Profiling
```
CUDA_VISIBLE_DEVICES=0,1,2,3 python3 -m sglang.launch_server --model-path openai/gpt-oss-120b --tp-size 4 --port 30001 

curl http://127.0.0.1:30001/flush_cache
python3 -m sglang.bench_serving --backend sglang-oai  --dataset-name random --random-input-len 512 --random-output-len 1024 --random-range-ratio 1 --num-prompts 20 --max-concurrency 1 --output-file 1.jsonl --port 30001 --warmup-requests 10
curl http://127.0.0.1:30001/flush_cache
python3 -m sglang.bench_serving --backend sglang-oai  --dataset-name random --random-input-len 512 --random-output-len 1024 --random-range-ratio 1 --num-prompts 200 --max-concurrency 32 --output-file 1.jsonl --port 30001 --warmup-requests 10
curl http://127.0.0.1:30001/flush_cache
python3 -m sglang.bench_serving --backend sglang-oai  --dataset-name random --random-input-len 512 --random-output-len 1024 --random-range-ratio 1 --num-prompts 300 --max-concurrency 64 --output-file 1.jsonl --port 30001 --warmup-requests 10
curl http://127.0.0.1:30001/flush_cache
python3 -m sglang.bench_serving --backend sglang-oai  --dataset-name random --random-input-len 512 --random-output-len 1024 --random-range-ratio 1 --num-prompts 400 --max-concurrency 128 --output-file 1.jsonl --port 30001 --warmup-requests 10

This PR:
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
|    |   max_concurrency |   input_throughput |   output_throughput |   mean_ttft_ms |   median_ttft_ms |   p99_ttft_ms |   mean_tpot_ms |   median_tpot_ms |   p99_tpot_ms |   per_user_throughput |
+====+===================+====================+=====================+================+==================+===============+================+==================+===============+=======================+
|  0 |             1.000 |            122.492 |             244.985 |         67.134 |           64.838 |       101.212 |          4.019 |            4.045 |         4.223 |               244.985 |
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
|  1 |            32.000 |           2228.420 |            4456.840 |        245.345 |          247.037 |       379.229 |          6.359 |            6.443 |         6.544 |               139.276 |
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
|  2 |            64.000 |           4153.943 |            8307.886 |        390.332 |          386.048 |       506.910 |          6.809 |            6.757 |         7.245 |               129.811 |
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
|  3 |           128.000 |           6205.168 |           12410.335 |        464.221 |          481.355 |       632.421 |          8.242 |            8.315 |         8.620 |                96.956 |
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+

Main:
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
|    |   max_concurrency |   input_throughput |   output_throughput |   mean_ttft_ms |   median_ttft_ms |   p99_ttft_ms |   mean_tpot_ms |   median_tpot_ms |   p99_tpot_ms |   per_user_throughput |
+====+===================+====================+=====================+================+==================+===============+================+==================+===============+=======================+
|  0 |             1.000 |            116.732 |             233.463 |         67.612 |           63.121 |       120.138 |          4.220 |            4.252 |         4.798 |               233.463 |
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
|  1 |            32.000 |           2147.463 |            4294.927 |        252.945 |          198.376 |       548.624 |          6.655 |            6.725 |         6.871 |               134.216 |
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
|  2 |            64.000 |           4085.339 |            8170.677 |        315.211 |          293.034 |       423.917 |          7.014 |            6.985 |         7.294 |               127.667 |
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
|  3 |           128.000 |           5996.815 |           11993.630 |        540.879 |          542.956 |       869.169 |          8.485 |            8.565 |         8.902 |                93.700 |
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+

```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: sgl-kernel/benchmark/bench_top_k_top_p_sampling.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        line_arg="provider",
+        line_vals=["torch", "sglang"],
+        line_names=["Torch Reference", "SGL Kernel"],
+        styles=[("red", "-"), ("green", "-")],
+        ylabel="us",
+        plot_name="top-k-top-p-joint-sampling-performance",
+        args={},
+    )
+)
+def benchmark_sampling(batch_size, vocab_size, p, provider):
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply sgl-kernel/benchmark/bench_top_k_top_p_sampling.py:90]: in TODO list.

File: python/sglang/srt/layers/sampler.py [hunk 2]
```diff
@@ -76,29 +76,34 @@ def forward(
...
-                            probs.contiguous(),
+                        logits.div_(sampling_info.temperatures)
+                        batch_next_token_ids = top_k_top_p_sampling_from_logits(
+                            logits,
                             sampling_info.top_ks,
                             sampling_info.top_ps,
-                            filter_apply_order="joint",
-                            check_nan=self.use_nan_detection,
+                            sampling_info.temperatures,
                         )
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/layers/sampler.py]: Fixed.

File: python/sglang/srt/layers/sampler.py [hunk 3]
```diff
@@ -76,29 +76,34 @@ def forward(
...
-            probs = logits
-            del logits
-
             if True:  # Keep this redundant check to simplify some internal code sync
                 if global_server_args_dict["sampling_backend"] == "flashinfer":
                     if sampling_info.need_min_p_sampling:
+                        logits.div_(sampling_info.temperatures)
+                        logits[:] = torch.softmax(logits, dim=-1)
+                        probs = logits
+                        del logits
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/layers/sampler.py]: This code dup is acceptable, as if sampling_info.need_min_p_sampling is False, the logic is different.

File: sgl-kernel/python/sgl_kernel/sampling.py [hunk 4]
```diff
@@ -383,3 +383,161 @@ def min_p_sampling_from_probs(
...
+            if torch.any(torch.isnan(probs)):
+                raise ValueError("Input probs contains NaN.")
+        return _top_k_top_p_sampling_from_probs_internal(
+            probs,
+            indices,
+            *_to_tensor_scalar_tuple(top_k),
+            *_to_tensor_scalar_tuple(top_p),
+            deterministic,
+            generator,
+        )
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply sgl-kernel/python/sgl_kernel/sampling.py:541]: This is for the consideration of performance implications.

**Review by BBuf — APPROVED** [review]
